typerep-map.cabal

```diff
@@ -42,8 +42,8 @@ library
                        Data.TypeRepMap.Internal
   ghc-options:         -Wall
   build-depends:       base >= 4.10 && < 5
-                     , containers
-                     , ghc-prim
+                     , containers >= 0.5.10.2
+                     , ghc-prim >= 0.5.1.1
                      , primitive >= 0.6.4
```

Comment:
It's not enough to specify only _lower bounds_ of dependencies, you also should specify _upper bounds_. For example, for `ghc-prim` it could be `ghc-prim >= 0.5.1.1 && < 0.6`. According to the PvP versioning policy, letters `A` and `B` in `A.B.C.D` version correspond to major releases of the package. Major releases can contain breaking change in the library API. So, in order to protect our packages from accidental breakage, we should add upper bounds as well.

Usually you want to specify bounds in the form of `foo >= x.y.z.t && < x.(y+1)`. Fortunately, there's special cabal syntax sugar for this. So if you write `ghc-prim ^>= 0.5.1.1` it will expand into `ghc-prim >= 0.5.1.1 && < 0.6`.

But we don't want this for all libraries. For example, there's already `0.6` version of `containers` package. So for `containers` package we want to specify constraints in the form like: `containers >= 0.5.10.2 && < 0.7`.

I hope this clarifies things about proper bounds management for packages!

Reply:
Thanks for the clarification. I have one question about the `containers` example though : since any version of `containers` above `0.6.x.y` would possibly introduce a breaking change, why do we want to set the upper bound to `0.7` instead of `0.6` (hence using `^>=`) ?

Reply:
>  why do we want to set the upper bound to 0.7 instead of 0.6

In this case we know for sure that `containers-0.6` won't break our particular code because we don't use any features that are subject of breakage.